File: 198.house-robber.cpp
/*
 * [198] House Robber
 *
 * https://leetcode.com/problems/house-robber
 *
 * Easy (38.69%)
 * Total Accepted:
 * Total Submissions:
 * Testcase Example:  '[]'
 *
 * You are a professional robber planning to rob houses along a street. Each
 * house has a certain amount of money stashed, the only constraint stopping
 * you from robbing each of them is that adjacent houses have security system
 * connected and it will automatically contact the police if two adjacent
 * houses were broken into on the same night.
 *
 * Given a list of non-negative integers representing the amount of money of
 * each house, determine the maximum amount of money you can rob tonight
 * without alerting the police.
 *
 * Credits:Special thanks to @ifanchu for adding this problem and creating all
 * test cases. Also thanks to @ts for adding additional test cases.
 */
class Solution {
public:
    /*
    dp[0] = nums[0];
    dp[1] = max(nums[0], nums[1]);
    dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])

    */
    int rob(vector<int>& nums) {
        int rob = 0;
        int notrob = 0;
        for(int i = 0; i < nums.size(); i++){
            int currob = notrob + nums[i];
            notrob = std::max(notrob, rob);
            rob = currob;
        }
        return std::max(rob, notrob);
    }
};

File: rust/src/leetcode/198.house-robber.rs
/*
 * @lc app=leetcode id=198 lang=rust
 *
 * [198] House Robber
 *
 * https://leetcode.com/problems/house-robber/description/
 *
 * algorithms
 * Medium (45.08%)
 * Likes:    9225
 * Dislikes: 223
 * Total Accepted:    880.7K
 * Total Submissions: 2M
 * Testcase Example:  '[1,2,3,1]'
 *
 * You are a professional robber planning to rob houses along a street. Each
 * house has a certain amount of money stashed, the only constraint stopping
 * you from robbing each of them is that adjacent houses have security systems
 * connected and it will automatically contact the police if two adjacent
 * houses were broken into on the same night.
 *
 * Given an integer array nums representing the amount of money of each house,
 * return the maximum amount of money you can rob tonight without alerting the
 * police.
 *
 *
 * Example 1:
 *
 *
 * Input: nums = [1,2,3,1]
 * Output: 4
 * Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
 * Total amount you can rob = 1 + 3 = 4.
 *
 *
 * Example 2:
 *
 *
 * Input: nums = [2,7,9,3,1]
 * Output: 12
 * Explanation: Rob house 1 (money = 2), rob house 3 (money = 9) and rob house
 * 5 (money = 1).
 * Total amount you can rob = 2 + 9 + 1 = 12.
 *
 *
 *
 * Constraints:
 *
 *
 * 1 <= nums.length <= 100
 * 0 <= nums[i] <= 400
 *
 *
 */

// @lc code=start
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        match nums.len() {
            0 => 0,
            1 => nums[0],
            _ => {
                let mut dp = vec![nums[0], nums[0].max(nums[1])];
                for i in 2..nums.len() {
                    dp.push((nums[i] + dp[i - 2]).max(dp[i - 1]));
                }
                dp.pop().unwrap()
            },
        }

    }
}
// @lc code=end
struct Solution;

#[cfg(test)]
#[test]
fn test() {
    assert_eq!(Solution::rob(vec![2,7,9,3,1]), 12);
}